File: core/python/usage_example.py
"""
Example usage of the new Python architecture
"""

from dynamic_filter_core.validation import Operator
from dynamic_filter_core.interfaces import FilterDefinition, FilterRequest, ContextAdapter, ConditionAdapterBuilder, Condition
from dynamic_filter_core.test_types import TestPropertyRef, TestPropertyRefImpl, get_property_ref

# Example 1: Creating a FilterDefinition
name_filter = FilterDefinition[TestPropertyRef](
    ref=TestPropertyRef.USER_NAME,
    operator=Operator.LIKE,
    value="John%"
)

age_filter = FilterDefinition[TestPropertyRef](
    ref=TestPropertyRef.USER_AGE,
    operator=Operator.GREATER_THAN,
    value=18
)

# Example 2: Creating a FilterRequest
filter_request = FilterRequest[TestPropertyRef](
    filters={
        "nameFilter": name_filter,
        "ageFilter": age_filter
    },
    combine_with="AND"
)

# Example 3: Type safety validation
def validate_filter(definition: FilterDefinition[TestPropertyRef]) -> None:
    """Validates that the filter definition is type-safe."""
    ref = definition.ref
    property_ref = get_property_ref(ref)
    
    if not property_ref.supports_operator(definition.operator):
        raise ValueError(
            f"Operator {definition.operator} not supported for property {ref}"
        )

# Test validation
try:
    validate_filter(name_filter)
    validate_filter(age_filter)
    print("✅ Python validation successful!")
except ValueError as e:
    print(f"❌ Python validation failed: {e}")

# Example 4: ContextAdapter usage
class ExampleContextAdapter(ContextAdapter[object, TestPropertyRef]):
    """Example implementation of ContextAdapter."""
    
    def __init__(self):
        self._conditions = {}
    
    def add_condition(self, filter_key: str, definition: FilterDefinition[TestPropertyRef]) -> None:
        """Adds a condition for the given filter key."""
        # Validate the filter
        validate_filter(definition)
        
        # Store the condition (simplified)
        self._conditions[filter_key] = definition
    
    def get_condition(self, filter_key: str) -> Condition:
        """Retrieves the condition associated with the given filter key."""
        return self._conditions.get(filter_key)

# Example usage
adapter = ExampleContextAdapter()
adapter.add_condition("nameFilter", name_filter)
adapter.add_condition("ageFilter", age_filter)

print("✅ Python ContextAdapter usage successful!")

if __name__ == "__main__":
    print("Python core architecture validation completed!")

File: core/csharp/UsageExample.cs
using System;
using System.Collections.Generic;
using DynamicFilter.Core.Models;
using DynamicFilter.Core.Validation;
using DynamicFilter.Core.TestTypes;
using DynamicFilter.Core.Interfaces;

namespace DynamicFilter.Core
{
    /// <summary>
    /// Example usage of the new C# architecture
    /// </summary>
    public class UsageExample
    {
        /// <summary>
        /// Demonstrates the new type-safe architecture
        /// </summary>
        public static void DemonstrateArchitecture()
        {
            // Example 1: Creating a FilterDefinition
            var nameFilter = new FilterDefinition<TestPropertyRef>(
                ref: TestPropertyRef.UserName,
                @operator: Operator.Like,
                value: "John%"
            );

            var ageFilter = new FilterDefinition<TestPropertyRef>(
                ref: TestPropertyRef.UserAge,
                @operator: Operator.GreaterThan,
                value: 18
            );

            // Example 2: Creating a FilterRequest
            var filterRequest = new FilterRequest<TestPropertyRef>(
                filters: new Dictionary<string, FilterDefinition<TestPropertyRef>>
                {
                    { "nameFilter", nameFilter },
                    { "ageFilter", ageFilter }
                },
                combineWith: "AND"
            );

            // Example 3: Type safety validation
            ValidateFilter(nameFilter);
            ValidateFilter(ageFilter);

            Console.WriteLine("✅ C# validation successful!");
        }

        /// <summary>
        /// Validates that the filter definition is type-safe
        /// </summary>
        /// <param name="definition">The filter definition to validate</param>
        private static void ValidateFilter(FilterDefinition<TestPropertyRef> definition)
        {
            var propertyRef = TestPropertyRefImpl.GetPropertyRef(definition.Ref);
            
            if (!propertyRef.SupportsOperator(definition.Operator))
            {
                throw new ArgumentException(
                    $"Operator {definition.Operator} not supported for property {definition.Ref}"
                );
            }
        }

        /// <summary>
        /// Example ContextAdapter usage
        /// </summary>
        public class ExampleContextAdapter : IContextAdapter<object, TestPropertyRef>
        {
            private readonly Dictionary<string, FilterDefinition<TestPropertyRef>> _conditions = new();

            public void AddCondition(string filterKey, FilterDefinition<TestPropertyRef> definition)
            {
                // Validate the filter
                ValidateFilter(definition);
                
                // Store the condition (simplified)
                _conditions[filterKey] = definition;
            }

            public ICondition? GetCondition(string filterKey)
            {
                return _conditions.TryGetValue(filterKey, out var definition) ? null : null; // Simplified
            }
        }

        /// <summary>
        /// Main entry point
        /// </summary>
        public static void Main()
        {
            try
            {
                DemonstrateArchitecture();
                Console.WriteLine("C# core architecture validation completed!");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"❌ C# validation failed: {ex.Message}");
            }
        }
    }
}

